sxg_rs/src/process_html.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::http::HttpResponse;
+use crate::http_parser::link::Link;
+use crate::link::ALLOWED_PARAM_NAMES;
+use anyhow::Result;
+use serde::Deserialize;
+use std::borrow::Cow;
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ProcessHtmlOption {
+    is_sxg: bool,
+}
+
+/// Processes HTML using the following processors.
+/// - For `<link rel=preload>` elements, they are promoted to Link headers.
+/// - For `<template data-sxg-only>` elements:
+///   - If SXG, they are "unwrapped" (i.e. their children promoted out of the <teplate>).
+///   - Else, they are deleted.
+/// - For `<script data-issxg-var>` elements, they are replaced with
+///   `<script>window.isSXG=...</script>`, where `...` is true or false.
+/// This function assumes the input document to be encoded in UTF8. It is needed to ensure that the
+/// document is explicitly labellbed as UTF8 via ContentType of <meta> before using this function.
+pub fn process_html(input: HttpResponse, option: ProcessHtmlOption) -> Result<HttpResponse> {
```

Comment:
I think you'll also want to port the code that ensures the content is explicitly labeled as UTF-8. That's in two pieces:

1. [Ensure there isn't a UTF-16 BOM.](https://github.com/google/sxg-rs/blob/79763737b635e0b0a1dd332b7ff6d974ab2a4597/typescript_utilities/src/processor.ts#L131-L139)
2. [Ensure there is a `<meta charset>` tag.](https://github.com/google/sxg-rs/blob/79763737b635e0b0a1dd332b7ff6d974ab2a4597/typescript_utilities/src/processor.ts#L154-L172)

This ensures that:
- The processors only run on applicable documents, so the measurement matches production.
- The processors don't corrupt up the document in ways that modify its LCP. (Per my testing, HTMLRewriter corrupted UTF-16 pages; I don't know if the same is true of lol-html.)
- We don't forget this step when porting to Rust-only environments like fastly_compute.

Reply:
Please review the new commits parsing `<meta charset>` tag.
We don't need an extra step to check BOM, becase `String::from_utf8` returns an error if the input contains UTF-16  BOM.